### PR TITLE
[6.2.z] CLI Errata, cherry-picked from 6.3 with some additions

### PR DESCRIPTION
**Pull request by ldjebran**
```bash
(2.7) dlezz@elysion:~/projects/robottelo-fork$ py.test tests/foreman/cli/test_errata.py -v -k "ErrataTestCase"
================================================= test session starts ==================================================
platform linux2 -- Python 2.7.12, pytest-3.0.3, py-1.4.31, pluggy-0.4.0 -- /home/dlezz/bin/redhat/python/2.7/bin/python
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile: 
plugins: cov-2.4.0, xdist-1.15.0
collected 34 items 

tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_affected_chosts <- ../../bin/redhat/python/2.7/lib/python2.7/site-packages/unittest2/case.py SKIPPED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_affected_chosts_by_erratum_restrict_flag <- ../../bin/redhat/python/2.7/lib/python2.7/site-packages/unittest2/case.py SKIPPED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_cve <- robottelo/decorators/__init__.py PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_org_id PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_org_id_and_sort_by_issued_date PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_org_id_and_sort_by_updated_date PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_org_label PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_org_label_and_sort_by_issued_date PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_org_label_and_sort_by_updated_date PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_org_name PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_org_name_and_sort_by_issued_date PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_org_name_and_sort_by_updated_date PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_product_id <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_product_id_and_org_id PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_product_id_and_org_label PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_product_id_and_org_name PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_product_name <- ../../bin/redhat/python/2.7/lib/python2.7/site-packages/unittest2/case.py SKIPPED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_product_name_and_org_id PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_product_name_and_org_label PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_filter_by_product_name_and_org_name PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_list_sort_by_issued_date PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_user_permission PASSED
tests/foreman/cli/test_errata.py::ErrataTestCase::test_positive_view_available_count_in_affected_chosts <- ../../bin/redhat/python/2.7/lib/python2.7/site-packages/unittest2/case.py SKIPPED

================================================= 11 tests deselected ==================================================
================================ 18 passed, 5 skipped, 11 deselected in 1429.30 seconds ================================
(2.7) dlezz@elysion:~/projects/robottelo-fork$ 
```
**PLEASE DO NOT SQUASH WHEN MERGING**

### CODE REVIEW

File: tests/foreman/cli/test_errata.py [hunk 1]
```diff
@@ -401,12 +910,66 @@ def test_positive_list_filter_by_product_id_and_org_id(self):
...
+        )
+        self.assertIn(self.org_product_errata_id, product_errata_ids)
+        self.assertIn(
+            self.org_multi_product_small_errata_id, product_small_errata_ids)
+        self.assertIn(
+            self.org_multi_product_big_errata_id, product_big_errata_ids)
+        self.assertEqual(
+            product_errata_ids.intersection(
+                product_small_errata_ids),
+            set([])
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: Any reason to pass empty list here?

```python
>>> set()
set([])
>>> set([])
set([])
```

File: tests/foreman/cli/test_errata.py [hunk 2]
```diff
@@ -401,12 +910,66 @@ def test_positive_list_filter_by_product_id_and_org_id(self):
...
+        )
+        self.assertEqual(
+            product_errata_ids.intersection(
+                product_big_errata_ids),
+            set([])
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: Same as previous comment

File: tests/foreman/cli/test_errata.py [hunk 3]
```diff
@@ -401,12 +910,66 @@ def test_positive_list_filter_by_product_id_and_org_id(self):
...
+        )
+        self.assertEqual(
+            product_small_errata_ids.intersection(
+                product_big_errata_ids),
+            set([])
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: Same as previous comment

File: tests/foreman/cli/test_errata.py [hunk 4]
```diff
@@ -419,12 +982,65 @@ def test_positive_list_filter_by_product_id_and_org_name(self):
...
+            self.org_multi_product_big_erratum_count
+        )
+        self.assertIn(self.org_product_errata_id, product_errata_ids)
+        self.assertIn(
+            self.org_multi_product_small_errata_id, product_small_errata_ids)
+        self.assertIn(
+            self.org_multi_product_big_errata_id, product_big_errata_ids)
+        self.assertEqual(
+            product_errata_ids.intersection(product_small_errata_ids),
+            set([])
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: Same as previous comment

File: tests/foreman/cli/test_errata.py [hunk 5]
```diff
@@ -419,12 +982,65 @@ def test_positive_list_filter_by_product_id_and_org_name(self):
...
+        )
+        self.assertEqual(
+            product_errata_ids.intersection(
+                product_big_errata_ids),
+            set([])
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: Same as previous comment

File: tests/foreman/cli/test_errata.py [hunk 6]
```diff
@@ -419,12 +982,65 @@ def test_positive_list_filter_by_product_id_and_org_name(self):
...
+        )
+        self.assertEqual(
+            product_small_errata_ids.intersection(
+                product_big_errata_ids),
+            set([])
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: Same as previous comment

File: tests/foreman/cli/test_errata.py [hunk 7]
```diff
@@ -438,10 +1054,61 @@ def test_positive_list_filter_by_product_id_and_org_label(self):
...
+            self.org_multi_product_big_erratum_count
+        )
+        self.assertIn(self.org_product_errata_id, product_errata_ids)
+        self.assertIn(
+            self.org_multi_product_small_errata_id, product_small_errata_ids)
+        self.assertIn(
+            self.org_multi_product_big_errata_id, product_big_errata_ids)
+        self.assertEqual(
+            product_errata_ids.intersection(product_small_errata_ids),
+            set([])
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: Same as previous comment

File: tests/foreman/cli/test_errata.py [hunk 8]
```diff
@@ -438,10 +1054,61 @@ def test_positive_list_filter_by_product_id_and_org_label(self):
...
+        )
+        self.assertEqual(
+            product_errata_ids.intersection(product_big_errata_ids),
+            set([])
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: Same as previous comment

File: tests/foreman/cli/test_errata.py [hunk 9]
```diff
@@ -438,10 +1054,61 @@ def test_positive_list_filter_by_product_id_and_org_label(self):
...
+        )
+        self.assertEqual(
+            product_small_errata_ids.intersection(product_big_errata_ids),
+            set([])
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: Same as previous comment

File: tests/foreman/cli/test_errata.py [hunk 10]
```diff
@@ -474,12 +1142,63 @@ def test_positive_list_filter_by_product_name_and_org_id(self):
         1. erratum list --product=<product_name> --organization-id=<org_id>
 
         @Assert: Errata is filtered by product name and Org id.
-
-        @caseautomation: notautomated
-
         """
-
-    @stubbed()
+        product_erratum_list = Erratum.list({
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: I compared the assert part of test bellow this line and it is exactly the same as the on begining on line 1085. I suppose it is the same on other tests that I made comment about set with empty list. So what you think about extraction a method `assert_errata_conditions' receiving the all errata related as parameters? At least we would have less code and no repetition.

File: tests/foreman/cli/test_errata.py [hunk 11]
```diff
@@ -492,12 +1211,63 @@ def test_positive_list_filter_by_product_name_and_org_name(self):
...
+        product_big_erratum_list = Erratum.list({
+            'organization': self.org_multi['name'],
+            'product': self.org_multi_product_big['name'],
+            'per-page': 1000
+        })
+        product_big_errata_ids = {
+            errata['errata-id']
+            for errata in product_big_erratum_list
+        }
+        self.assertEqual(
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: It seems same code as previous comment.

File: tests/foreman/cli/test_errata.py [hunk 12]
```diff
@@ -511,12 +1281,63 @@ def test_positive_list_filter_by_product_name_and_org_label(self):
...
+        product_big_erratum_list = Erratum.list({
+            'organization-label': self.org_multi['label'],
+            'product': self.org_multi_product_big['name'],
+            'per-page': 1000
+        })
+        product_big_errata_ids = {
+            errata['errata-id']
+            for errata in product_big_erratum_list
+        }
+        self.assertEqual(
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: It seems same code as previous comment.

File: tests/foreman/cli/test_errata.py [hunk 13]
```diff
@@ -529,12 +1350,34 @@ def test_positive_list_filter_by_org_id(self):
...
+            for errata in errata_org_erratum_list
+        }
+        self.assertEqual(len(org_errata_ids), self.org_product_erratum_count)
+        self.assertEqual(
+            len(errata_org_errata_ids), self.errata_org_product_erratum_count)
+        self.assertIn(self.org_product_errata_id, org_errata_ids)
+        self.assertIn(self.errata_org_product_errata_id, errata_org_errata_ids)
+        self.assertEqual(
+            org_errata_ids.intersection(errata_org_errata_ids),
+            set([])
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: set with empty list

File: tests/foreman/cli/test_errata.py [hunk 14]
```diff
@@ -547,12 +1390,34 @@ def test_positive_list_filter_by_org_name(self):
...
+        }
+        errata_org_erratum_list = Erratum.list({
+            'organization': self.errata_org['name'],
+            'per-page': 1000
+        })
+        errata_org_errata_ids = {
+            errata['errata-id']
+            for errata in errata_org_erratum_list
+        }
+        self.assertEqual(len(org_errata_ids), self.org_product_erratum_count)
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: It seem same code as the one starting on line 1370. Same method extraction could apply on this code, similar to the one I proposed.

File: tests/foreman/cli/test_errata.py [hunk 15]
```diff
@@ -565,12 +1430,36 @@ def test_positive_list_filter_by_org_label(self):
...
+        }
+        errata_org_erratum_list = Erratum.list({
+            'organization-label': self.errata_org['label'],
+            'per-page': 1000
+        })
+        errata_org_errata_ids = {
+            errata['errata-id']
+            for errata in errata_org_erratum_list
+        }
+        self.assertEqual(len(org_errata_ids), self.org_product_erratum_count)
```
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: It seem same code as the one starting on line 1370. Same method extraction could apply on this code, similar to the one I proposed.

File: tests/foreman/cli/test_errata.py (hunk 1 above)
Comment (renzon) [inline tests/foreman/cli/test_errata.py]: I mentioned a lot of cases that is using this code. But if decide extract a method with common code, you could fix the code only on this extracted method.

File: tests/foreman/cli/test_errata.py (hunk 15 above)
Comment (ldjebran) [inline tests/foreman/cli/test_errata.py]: thanks @renzon for your comments, the primary purpose of this PR is to transfer the tests from 6.3 to 6.2 and to add host collection tests that cannot be tested in 6.3 because of a bug, this module has been refactored already, but for the moment any refactoring should be done in a separate PR  in 6.3 and after than cherry-picked back to 6.2.z

Comment (renzon) [inline tests/foreman/cli/test_errata.py]: Got it

**Comment by renzon** [comment]
ACK
